src/components/tl/ucp/scatterv/scatterv_linear.c

```diff
@@ -0,0 +1,123 @@
+    ucc_rank_t         grank = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t         gsize = UCC_TL_TEAM_SIZE(team);
+    int                polls = 0;
+    ucc_rank_t         peer, posts, nreqs;
+    size_t             dt_size, data_size, data_displ;
+
+    if (UCC_IS_ROOT(*args, grank)) {
+        posts   = UCC_TL_UCP_TEAM_LIB(team)->cfg.scatterv_linear_num_posts;
+        nreqs   = (posts > gsize || posts == 0) ? gsize : posts;
+        dt_size = ucc_dt_size(TASK_ARGS(task).src.info_v.datatype);
+
+        while (polls++ < task->n_polls) {
+            while ((task->tagged.send_posted < gsize) &&
+                   ((task->tagged.send_posted - task->tagged.send_completed) <
+                    nreqs)) {
+                peer       = get_peer(grank, gsize, task->tagged.send_posted);
+                data_size  = ucc_coll_args_get_count(args,
+                                args->src.info_v.counts, peer) * dt_size;
+                data_displ = ucc_coll_args_get_displacement(args,
+                                args->src.info_v.displacements, peer) * dt_size;
+                UCPCHECK_GOTO(ucc_tl_ucp_send_nz(PTR_OFFSET(sbuf, data_displ),
+                                                 data_size, smem, peer, team, task),
+                               task, out);
+                polls = 0;
+            }
+        }
+        if (task->tagged.send_posted < gsize) {
+            return;
+        }
+    }
+
+    task->super.status = ucc_tl_ucp_test(task);
+out:
+    if (task->super.status != UCC_INPROGRESS) {
+        UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_scatterv_linear_done",
+                                         0);
+    }
+}
+
+ucc_status_t ucc_tl_ucp_scatterv_linear_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_ucp_task_t *task  = ucc_derived_of(coll_task, ucc_tl_ucp_task_t);
+    ucc_tl_ucp_team_t *team  = TASK_TEAM(task);
+    ucc_coll_args_t   *args  = &TASK_ARGS(task);
+    ucc_rank_t         grank = UCC_TL_TEAM_RANK(team);
+    ucc_memory_type_t  smem  = args->src.info_v.mem_type;
+    ucc_memory_type_t  rmem  = args->dst.info.mem_type;
+    void              *rbuf  = args->dst.info.buffer;
+    void *             sbuf;
+    size_t             dt_size, data_displ, data_size;
+
+    UCC_TL_UCP_PROFILE_REQUEST_EVENT(coll_task, "ucp_scatterv_linear_start", 0);
+    ucc_tl_ucp_task_reset(task, UCC_INPROGRESS);
+
+    if (UCC_IS_ROOT(*args, grank)) {
+        if (!UCC_IS_INPLACE(*args)) {
+            dt_size    = ucc_dt_size(args->src.info_v.datatype);
+            data_size  = ucc_coll_args_get_count(args,
+                            args->src.info_v.counts, grank) * dt_size;
+            data_displ = ucc_coll_args_get_displacement(args,
+                            args->src.info_v.displacements, grank) * dt_size;
+            sbuf       = PTR_OFFSET(args->src.info_v.buffer, data_displ);
+
+            UCPCHECK_GOTO(ucc_tl_ucp_recv_nz(rbuf, data_size, rmem, grank, team,
```

Comment:
Why are these (send & recv) in start instead of progress? Can be moved with flag to make sure root "goes" first.

Reply:
root rank recv data only once, if move this recv to progress will need to check if recv posted every time we enter to progress function.